src/services/local-storage/useLocalStorage.ts

```diff
@@ -1,32 +1,74 @@
-import type { Dispatch, SetStateAction } from 'react'
-import { useState, useCallback, useEffect } from 'react'
+import { useCallback, useEffect, useState } from 'react'
 import local from './local'
 
-const useLocalStorage = <T>(key: string, initialState: T): [T, Dispatch<SetStateAction<T>>] => {
-  const [cache, setCache] = useState<T>(initialState)
+// The setter accepts T or a function that takes the old value and returns T
+// Mimics the behavior of useState
+type Undefinable<T> = T | undefined
 
-  useEffect(() => {
-    const initialValue = local.getItem<T>(key)
-    if (initialValue !== undefined) {
-      setCache(initialValue)
-    }
-  }, [setCache, key])
+type Setter<T> = (val: T | ((prevVal: Undefinable<T>) => Undefinable<T>)) => void
+
+const useLocalStorage = <T>(key: string): [Undefinable<T>, Setter<T>] => {
+  const [cache, setCache] = useState<Undefinable<T>>()
+
+  // This is the setter that will be returned
+  // It will update the local storage and cache
+  const setNewValue = useCallback<Setter<T>>(
+    (value) => {
+      setCache((oldValue) => {
+        const newValue = value instanceof Function ? value(oldValue) : value
 
-  const setNewValue = useCallback(
-    (value: T | ((prevState: T) => T)) => {
-      setCache((prevState) => {
-        const newState = value instanceof Function ? value(prevState) : value
+        if (newValue !== oldValue) {
+          local.setItem(key, newValue)
 
-        if (newState !== prevState) {
-          local.setItem(key, newState)
+          // Dispatch a fake storage event within the current browser tab
+          // The real storage event is dispatched only in other tabs
+          window.dispatchEvent(
+            new StorageEvent('storage', {
+              key: local.getPrefixedKey(key),
+            }),
+          )
         }
 
-        return newState
+        return newValue
       })
     },
-    [setCache, key],
+    [key],
   )
 
+  // On mount, sync the cache with the local storage
+  useEffect(() => {
+    const lsValue = local.getItem<T>(key)
+    if (lsValue !== null) {
+      setCache(lsValue)
+    }
+  }, [key])
+
+  // Subscribe to changes in local storage and update the cache
+  // This will work across tabs
+  useEffect(() => {
+    const syncCache = () => {
+      const lsValue = local.getItem<T>(key)
+      if (lsValue !== null) {
+        setCache(lsValue)
+      }
+    }
+
+    const onStorageEvent = (event: StorageEvent) => {
+      if (event.key === local.getPrefixedKey(key)) {
+        syncCache()
+      }
+    }
+
+    // Set the initial value
+    syncCache()
```

Comment:
This seems to do the same as the other useEffect that sets the initial value on mount. Can we remove one or the other?